File: Test/clik-mcp/search-service/chunker.py
"""
Structure-Aware Recursive Chunker for Korean Council Minutes (Python port).

Splits pre-processed council minutes text (from ClikClient.stripHtml()) into
semantically coherent chunks that preserve:
  - Speaker attribution (◆ markers)
  - Agenda context (【...】 markers)
  - Sentence/paragraph boundaries
"""

from __future__ import annotations

import re
from dataclasses import dataclass, field
from typing import Optional


@dataclass
class Chunk:
    """A single chunk of council minutes text with metadata."""
    text: str
    speaker: Optional[str] = None
    agenda_context: Optional[str] = None
    start_idx: int = 0
    end_idx: int = 0
    index: int = 0


# ── Separator patterns ─────────────────────────────────────────────────
SPEAKER_RE = re.compile(r"◆\s*(.+)")
AGENDA_RE = re.compile(r"【([^】]*?)】?")

# Split hierarchy (applied in order of priority)
PARAGRAPH_SEP = re.compile(r"\n\n")
SENTENCE_SEP = re.compile(r"(?<=[.?!。])\s+")
CLAUSE_SEP = re.compile(r"(?<=[,，])\s+")

# Top-level structural boundary (speaker or agenda change)
STRUCTURE_BOUNDARY = re.compile(r"\n\n(?=◆ |【)")


def chunk_transcript(
    text: str,
    max_chunk_size: int = 1200,
    min_chunk_size: int = 100,
) -> list[Chunk]:
    """
    Chunk a council minutes transcript using structure-aware recursive splitting.

    Args:
        text: Pre-processed transcript text (output of stripHtml).
        max_chunk_size: Maximum characters per chunk (~800 tokens for Korean).
        min_chunk_size: Minimum characters; smaller fragments are discarded.

    Returns:
        List of Chunk objects with speaker/agenda metadata.
    """
    if not text or not text.strip():
        return []

    # Step 1: Split by top-level structural boundaries (speaker/agenda)
    raw_segments = _split_by_structure(text)

    # Step 2: Recursively split oversized segments
    chunks: list[Chunk] = []
    current_speaker: Optional[str] = None
    current_agenda: Optional[str] = None
    chunk_index = 0

    for seg_text, seg_start in raw_segments:
        # Update speaker/agenda context
        detected_speaker = _extract_speaker(seg_text)
        if detected_speaker:
            current_speaker = detected_speaker

        detected_agenda = _extract_agenda(seg_text)
        if detected_agenda:
            current_agenda = detected_agenda

        # Recursively split if too large
        sub_texts = _recursive_split(
            seg_text,
            max_chunk_size,
            [PARAGRAPH_SEP, SENTENCE_SEP, CLAUSE_SEP],
        )

        for sub in sub_texts:
            trimmed = sub.strip()
            if len(trimmed) < min_chunk_size:
                continue

            # Prepend speaker prefix if chunk doesn't start with one
            final_text = trimmed
            if current_speaker and not trimmed.startswith("◆"):
                final_text = f"[화자: {current_speaker}]\n{trimmed}"

            chunks.append(Chunk(
                text=final_text,
                speaker=current_speaker,
                agenda_context=current_agenda,
                start_idx=seg_start,
                end_idx=seg_start + len(sub),
                index=chunk_index,
            ))
            chunk_index += 1

    return chunks


# ── Internal helpers ───────────────────────────────────────────────────

def _split_by_structure(text: str) -> list[tuple[str, int]]:
    """Split text by speaker/agenda markers. Returns (text, start_idx) tuples."""
    segments: list[tuple[str, int]] = []
    last_idx = 0

    for match in STRUCTURE_BOUNDARY.finditer(text):
        seg = text[last_idx:match.start()]
        if seg.strip():
            segments.append((seg, last_idx))
        last_idx = match.start() + 2  # skip \n\n, keep marker

    remaining = text[last_idx:]
    if remaining.strip():
        segments.append((remaining, last_idx))

    return segments


def _recursive_split(
    text: str,
    max_size: int,
    separators: list[re.Pattern],
) -> list[str]:
    """Recursively split text using a hierarchy of separators."""
    if len(text) <= max_size:
        return [text]

    if not separators:
        return _hard_split(text, max_size)

    current_sep, *remaining_seps = separators
    pieces = current_sep.split(text)

    if len(pieces) <= 1:
        return _recursive_split(text, max_size, remaining_seps)

    result: list[str] = []
    buffer = ""

    for piece in pieces:
        if len(buffer) + len(piece) <= max_size:
            buffer += ("\n\n" if buffer else "") + piece
        else:
            if buffer:
                result.append(buffer)
            if len(piece) > max_size:
                result.extend(_recursive_split(piece, max_size, remaining_seps))
                buffer = ""
            else:
                buffer = piece

    if buffer:
        result.append(buffer)

    return result


def _hard_split(text: str, max_size: int) -> list[str]:
    """Last-resort: break at nearest whitespace before max_size."""
    pieces: list[str] = []
    remaining = text

    while len(remaining) > max_size:
        split_at = remaining.rfind(" ", 0, max_size)
        if split_at <= 0:
            split_at = remaining.rfind("\n", 0, max_size)
        if split_at <= 0:
            split_at = max_size

        pieces.append(remaining[:split_at])
        remaining = remaining[split_at:].lstrip()

    if remaining:
        pieces.append(remaining)

    return pieces


def _extract_speaker(text: str) -> Optional[str]:
    """Extract speaker name from ◆ marker."""
    m = SPEAKER_RE.search(text)
    if not m:
        return None
    name = m.group(1).split("\n")[0].strip().rstrip(": ")
    return name or None


def _extract_agenda(text: str) -> Optional[str]:
    """Extract agenda context from 【...】 marker."""
    m = AGENDA_RE.search(text)
    if not m:
        return None
    return m.group(1).strip() or None

File: Test/clik-mcp/src/chunker.ts
/**
 * Structure-Aware Recursive Chunker for Korean Council Minutes
 *
 * Replaces the naive 3,000-char fixed windowing with intelligent chunking
 * that preserves:
 *   - Speaker attribution (◆ markers from stripHtml())
 *   - Agenda context (【...】 markers)
 *   - Sentence/paragraph boundaries
 *
 * Designed for the already-stripped text output of ClikClient.stripHtml().
 */

export interface Chunk {
    /** The chunk text content */
    text: string;
    /** Speaker name for this chunk (carried forward if split mid-speech) */
    speaker: string | null;
    /** Current agenda item context */
    agendaContext: string | null;
    /** Original start index in the source text */
    startIdx: number;
    /** Original end index in the source text */
    endIdx: number;
    /** Chunk index in sequence */
    index: number;
}

export interface ChunkerOptions {
    /**
     * Target maximum characters per chunk.
     * Korean ≈ 1.5-3 tokens/char, so 1200 chars ≈ ~600-800 tokens.
     * @default 1200
     */
    maxChunkSize?: number;
    /**
     * Minimum chunk size to avoid tiny fragments.
     * @default 100
     */
    minChunkSize?: number;
}

// ── Separator hierarchy for Korean council minutes ─────────────────────
// Priority 1: Speaker change (◆ marker from stripHtml)
const SPEAKER_SEPARATOR = /\n\n◆ /;
// Priority 2: Agenda item (【 marker from stripHtml)
const AGENDA_SEPARATOR = /\n\n【/;
// Priority 3: Paragraph break
const PARAGRAPH_SEPARATOR = /\n\n/;
// Priority 4: Sentence-ending punctuation (Korean period, question/exclamation marks)
const SENTENCE_SEPARATOR = /(?<=[.?!。])\s+/;
// Priority 5: Comma or natural pause
const CLAUSE_SEPARATOR = /(?<=[,，])\s+/;

// Speaker name extraction: ◆ at line start, followed by name
const SPEAKER_REGEX = /◆\s*(.+)/;
// Agenda extraction: 【...】 or 【... (text until end of line)
const AGENDA_REGEX = /【([^】]*?)】?/;

/**
 * Chunk a council minutes transcript using structure-aware recursive splitting.
 *
 * The text is expected to be pre-processed by ClikClient.stripHtml(), which
 * produces markers like:
 *   ◆ 김의원  (speaker)
 *   【안건명】 (agenda item)
 *   ⏰ 시간    (timestamp)
 */
export function chunkTranscript(
    text: string,
    options: ChunkerOptions = {}
): Chunk[] {
    const { maxChunkSize = 1200, minChunkSize = 100 } = options;

    if (!text || text.trim().length === 0) {
        return [];
    }

    // Step 1: Split into top-level segments by speaker/agenda boundaries
    const rawSegments = splitByStructure(text);

    // Step 2: Recursively split oversized segments
    const chunks: Chunk[] = [];
    let currentSpeaker: string | null = null;
    let currentAgenda: string | null = null;
    let chunkIndex = 0;

    for (const segment of rawSegments) {
        // Update speaker/agenda context from this segment
        const detectedSpeaker = extractSpeaker(segment.text);
        if (detectedSpeaker) {
            currentSpeaker = detectedSpeaker;
        }

        const detectedAgenda = extractAgenda(segment.text);
        if (detectedAgenda) {
            currentAgenda = detectedAgenda;
        }

        // Recursively split if too large
        const subTexts = recursiveSplit(
            segment.text,
            maxChunkSize,
            [PARAGRAPH_SEPARATOR, SENTENCE_SEPARATOR, CLAUSE_SEPARATOR]
        );

        for (const subText of subTexts) {
            const trimmed = subText.trim();
            if (trimmed.length < minChunkSize) continue;

            // Prepend speaker prefix if chunk doesn't start with one
            let finalText = trimmed;
            if (currentSpeaker && !trimmed.startsWith("◆")) {
                finalText = `[화자: ${currentSpeaker}]\n${trimmed}`;
            }

            chunks.push({
                text: finalText,
                speaker: currentSpeaker,
                agendaContext: currentAgenda,
                startIdx: segment.startIdx,
                endIdx: segment.startIdx + subText.length,
                index: chunkIndex++,
            });
        }
    }

    return chunks;
}

// ── Internal helpers ───────────────────────────────────────────────────

interface RawSegment {
    text: string;
    startIdx: number;
}

/**
 * Split text by speaker and agenda markers (top-level structural boundaries).
 * Preserves the marker in each segment.
 */
function splitByStructure(text: string): RawSegment[] {
    // Combined pattern: split on speaker OR agenda boundaries
    const boundaryPattern = /\n\n(?=◆ |【)/g;
    const segments: RawSegment[] = [];

    let lastIndex = 0;
    let match: RegExpExecArray | null;

    while ((match = boundaryPattern.exec(text)) !== null) {
        const segText = text.slice(lastIndex, match.index);
        if (segText.trim().length > 0) {
            segments.push({ text: segText, startIdx: lastIndex });
        }
        lastIndex = match.index + 2; // skip the \n\n, keep the marker
    }

    // Last segment
    const remaining = text.slice(lastIndex);
    if (remaining.trim().length > 0) {
        segments.push({ text: remaining, startIdx: lastIndex });
    }

    return segments;
}

/**
 * Recursively split text using a hierarchy of separators.
 * Tries the first separator; if any resulting piece is still too large,
 * recursively tries the next separator on that piece.
 */
function recursiveSplit(
    text: string,
    maxSize: number,
    separators: RegExp[]
): string[] {
    if (text.length <= maxSize) {
        return [text];
    }

    if (separators.length === 0) {
        // No more separators — hard-split by maxSize at nearest space
        return hardSplit(text, maxSize);
    }

    const [currentSep, ...remainingSeps] = separators;
    const pieces = text.split(currentSep);

    if (pieces.length <= 1) {
        // Separator didn't split — try next level
        return recursiveSplit(text, maxSize, remainingSeps);
    }

    const result: string[] = [];
    let buffer = "";

    for (const piece of pieces) {
        if (buffer.length + piece.length <= maxSize) {
            buffer += (buffer ? "\n\n" : "") + piece;
        } else {
            if (buffer) result.push(buffer);
            if (piece.length > maxSize) {
                // This piece alone is too large — recurse deeper
                result.push(...recursiveSplit(piece, maxSize, remainingSeps));
                buffer = "";
            } else {
                buffer = piece;
            }
        }
    }

    if (buffer) result.push(buffer);

    return result;
}

/**
 * Last-resort hard split: break at the nearest whitespace before maxSize.
 */
function hardSplit(text: string, maxSize: number): string[] {
    const pieces: string[] = [];
    let remaining = text;

    while (remaining.length > maxSize) {
        // Find last space within maxSize boundary
        let splitAt = remaining.lastIndexOf(" ", maxSize);
        if (splitAt <= 0) {
            splitAt = remaining.lastIndexOf("\n", maxSize);
        }
        if (splitAt <= 0) {
            // No good break point — force split
            splitAt = maxSize;
        }

        pieces.push(remaining.slice(0, splitAt));
        remaining = remaining.slice(splitAt).trimStart();
    }

    if (remaining.length > 0) {
        pieces.push(remaining);
    }

    return pieces;
}

/** Extract speaker name from a text block (looks for ◆ marker) */
function extractSpeaker(text: string): string | null {
    const match = text.match(SPEAKER_REGEX);
    if (!match) return null;

    // Clean up: speaker name is typically one line
    const name = match[1].split("\n")[0].trim();
    // Remove trailing punctuation or whitespace artifacts
    return name.replace(/[:\s]+$/, "") || null;
}

/** Extract agenda context from a text block (looks for 【 marker) */
function extractAgenda(text: string): string | null {
    const match = text.match(AGENDA_REGEX);
    if (!match) return null;
    return match[1].trim() || null;
}
